### PR TITLE
Add llama3-8b context parallelism (ring attention)

### PR DESCRIPTION
**Pull request by sfvaroglu**
📝 Summary of Changes
This PR adds HLO for llama3-8b context parallelism (ring attention) to xla/tools/benchmarks/hlo.

🎯 Justification
We would like to track performance for this model.

🚀 Kind of Contribution
🧪 Tests

📊 Benchmark (for Performance Improvements)
On B200, average step time:
- NCCL: 0.7s
- NCCL + NCCL_NVLS_ENABLE=1 + --xla_gpu_enable_nccl_user_buffers=true: 0.58s

🧪 Unit Tests:
N/A.

🧪 Execution Tests:
N/A

### CODE REVIEW

**Comment by sfvaroglu** [comment]
@frgossen This is the HLO for llama3-8b context parallelism, which I shared results for last week. It's after the SPMD partitioner, without any metadata. Please let me know if anything else is needed. Thanks!

**Review by frgossen — APPROVED** [review]
Thank you!

**Comment by frgossen** [comment]
Can you rebase this PR?

**Comment by dimitar-asenov** [comment]
Hi @sfvaroglu. The file you're trying to add is larger than our 9MB limit for HLOs. Please reduce its size so that it fits under the limit.

**Comment by sfvaroglu** [comment]
> Hi @sfvaroglu. The file you're trying to add is larger than our 9MB limit for HLOs. Please reduce its size so that it fits under the limit.

Hi @dimitar-asenov, are there any specific recommendations or best practices you suggest for reducing the HLO file size to fit under the 9MB limit? This is from a run with llama3-8b context parallelism enabled.

**Comment by dimitar-asenov** [comment]
Looking at the HLO I see at least a few fairly mechanical things that can be reduced:
- Unify redundant computations (e.g. regions that do `add`).
- Drop % from all names
- Drop optional in/out spec for regions.
- Drop layouts that are already normalized.
- Drop arg types and shapes in instructions. Just the arg name is enough.

I think the above should be enough to get the file under the limit and they fully preserve the semantics of the HLO.

**Comment by akuegel** [comment]
> Looking at the HLO I see at least a few fairly mechanical things that can be reduced:
> 
> * Unify redundant computations (e.g. regions that do `add`).
> * Drop % from all names
> * Drop optional in/out spec for regions.
> * Drop layouts that are already normalized.
> * Drop arg types and shapes in instructions. Just the arg name is enough.
> 
> I think the above should be enough to get the file under the limit and they fully preserve the semantics of the HLO.

Some hints what tools can be used for this:

- To unify computations, we have https://github.com/openxla/xla/blob/main/xla/hlo/transforms/simplifiers/hlo_computation_deduplicator.h
- You can read the HLO, and write it again using other print options: https://github.com/openxla/xla/blob/main/xla/hlo/ir/hlo_print_options.h

**Comment by frgossen** [comment]
I think this is a benchmark generated from Maxtext. Updating this manually is probably not the right thing to do. I can add the benchmark as a file-based benchmark. Do you mind sharing the maxtext configuration for this? I can add that instead of the HLO.

**Comment by sfvaroglu** [comment]
> I think this is a benchmark generated from Maxtext. Updating this manually is probably not the right thing to do. I can add the benchmark as a file-based benchmark. Do you mind sharing the maxtext configuration for this? I can add that instead of the HLO.

Here is the cmd:
`python -m MaxText.train MaxText/configs/base.yml run_name=l3-8b-mprocess-cp8-fsdp1-BF16-ring-load_balanced-nccl scan_layers=false max_target_length=8192 attention=cudnn_flash_te steps=50 per_device_batch_size=1 model_name=llama3-8b remat_policy=minimal_flash enable_checkpointing=false logits_dot_in_fp32=false use_iota_embed=false tokenizer_path=assets/tokenizer_llama3.tiktoken tokenizer_type=tiktoken base_output_directory=../outputs/ profiler=nsys skip_first_n_steps_for_profiler=25 profiler_steps=3 quantization="" enable_goodput_recording=false monitor_goodput=false dataset_type=synthetic ici_fsdp_parallelism=1 ici_context_parallelism=8 cp_strategy=ring context_parallel_load_balance=true hardware=gpu_multiprocess`

This requires TE 2.7 (not released yet) + some MaxText changes.

**Comment by frgossen** [comment]
> This requires TE 2.7 (not released yet) + some MaxText changes.

I assume this means I cannot run this model yet, even if I did put in the effort of importing it?

**Comment by toli-y** [comment]
We need a model or HLO to benchmark this improvement.

**Review by toli-y — COMMENTED** [review]
How do we benchmark the improvement or verify that optimization works?

**Comment by sfvaroglu** [comment]
> How do we benchmark the improvement or verify that optimization works?

The MaxText changes haven't been upstreamed yet. Once that's done, we could add [this config](https://github.com/openxla/xla/pull/31108#issuecomment-3330189167) to the spreadsheet.

**Review by toli-y — APPROVED** [review]

**Comment by golechwierowicz** [comment]
Hey, the file is still above 9MB limit. Can you please apply suggestions [from Adrian's comment](https://github.com/openxla/xla/pull/31108#issuecomment-3317699758)?

**Comment by sfvaroglu** [comment]
> Hey, the file is still above 9MB limit. Can you please apply suggestions [from Adrian's comment](https://github.com/openxla/xla/pull/31108#issuecomment-3317699758)?

Please see [Frederik's comment](https://github.com/openxla/xla/pull/31108#issuecomment-3330110097). Perhaps we should close this PR if the "add the benchmark as a file-based benchmark" part is done.

**Comment by sfvaroglu** [comment]
> Please see [Frederik's comment](https://github.com/openxla/xla/pull/31108#issuecomment-3330110097). Perhaps we should close this PR if the "add the benchmark as a file-based benchmark" part is done.

The MaxText [PR](https://github.com/AI-Hypercomputer/maxtext/pull/2600) for this config has been merged. @felixwqp could you add this config to the spreadsheet tracker (since the HLO is too big)?

**Comment by golechwierowicz** [comment]
@sfvaroglu does it mean we can close this PR?

**Comment by sfvaroglu** [comment]
> @sfvaroglu does it mean we can close this PR?

Pending "@felixwqp could you add this config to the spreadsheet tracker (since the HLO is too big)?"

**Comment by felixwqp** [comment]
Clarifying internally about ` file-based benchmark` in https://github.com/openxla/xla/pull/31108#issuecomment-3330110097. will keep you posted once it's ready to merge.

**Comment by sfvaroglu** [comment]
> Clarifying internally about ` file-based benchmark` in [#31108 (comment)](https://github.com/openxla/xla/pull/31108#issuecomment-3330110097). will keep you posted once it's ready to merge.

@felixwqp The HLO is much smaller now (used scan_layers=true and 4-node config). Please review.

**Comment by felixwqp** [comment]
due to priority adjustment, pifon2a@(or his delegate) will collaborate  on the PR review as reviewer going forward. 

Feel free to ping me if any additional information is needed.

**Comment by golechwierowicz** [comment]
@pifon2a do we still need this?

**Review by pifon2a — APPROVED** [review]
